File: leetcode solutions/Python/subsets/subsets.py

# bit manipulation
class Solution(object):
    def subsets(self, nums):
        nums.sort()
        total = 2 ** len(nums)
        res = []

        for i in xrange(total):
            res.append([])

        for i in xrange(len(nums)):
            for j in xrange(total):
                if ((j>>i)&1) > 0:
                    res[j].append(nums[i])

        return res


# recursion
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        return self.subsetsHelper([], sorted(nums))


    def subsetsHelper(self, curr, nums):
        if not nums:
            return [curr]

        return self.subsetsHelper(curr, nums[1:]) + self.subsetsHelper(curr + [nums[0]], nums[1:])

# iteration
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = [[]]

        for i in xrange(len(nums)):
            size = len(result)
            for j in xrange(size):
                 result.append(list(result[j]))
                 result[-1].append(nums[i])
        return result

File: leetcode solutions/Java/subsets/subsets.java
public class Solution {

public List<List<Integer>> subsets(int[] S) {
    List<List<Integer>> result = new ArrayList<List<Integer>>();

    if(S.length == 0){
        return result;
    }

    Arrays.sort(S);
    dfs(S, 0, new ArrayList<Integer>(), result);
    return result;
}

public void dfs(int[] s, int index, List<Integer> path, List<List<Integer>> result){
    result.add(new ArrayList<Integer>(path));

    for(int i = index; i < s.length; i++){
        path.add(s[i]);
        dfs(s, i+1, path, result);
        path.remove(path.size()-1);
    }
  }
}